### PR TITLE
refactor(guest): replace external chown with syscalls

### PR DESCRIPTION
**Pull request by ltstriker**
## Summary

Replace the guest's PATH-resolved `chown -R` subprocess with an in-process, fd-relative syscall implementation while preserving the existing ownership sampling and best-effort repair behavior.

## Call graph

Before
  mount                     (BlockDeviceMount · src/guest/src/storage/block_device.rs:25) — mounts and initializes the guest filesystem
    └─ fix_if_needed        (OwnershipFixer · src/guest/src/storage/perms.rs:17) — samples ownership before repair
         ├─ ownership_matches (OwnershipFixer · src/guest/src/storage/perms.rs:74) — skips an already-matching tree
         └─ Command::new("chown") (std::process::Command · src/guest/src/storage/perms.rs:43) — runs PATH-resolved `chown -R`

After
  mount                     (BlockDeviceMount · src/guest/src/storage/block_device.rs:25) — mounts and initializes the guest filesystem
    └─ fix_if_needed        (OwnershipFixer · src/guest/src/storage/perms.rs:23) — preserves the existing ownership sample gate
         ├─ ownership_matches (OwnershipFixer · src/guest/src/storage/perms.rs:82) — keeps the original root-plus-five sampling policy
         └─ chown_path      (RecursiveChowner · src/guest/src/storage/perms.rs:134) — walks the tree iteratively with fd-relative syscalls
              ├─ chown_operand   (RecursiveChowner · src/guest/src/storage/perms.rs:316) — changes a non-directory root without following symlinks
              ├─ chown_entry     (RecursiveChowner · src/guest/src/storage/perms.rs:329) — changes descendant entries with `fchownat`
              └─ chown_directory (RecursiveChowner · src/guest/src/storage/perms.rs:342) — changes directories post-order with `fchown`

## Changes

- Remove the runtime `chown` binary, PATH lookup, and reaper-fence dependency from guest ownership repair.
- Add an iterative, fd-relative walker that handles non-UTF-8 names, symlinks, special files, cross-mount traversal, ancestor cycles, and bounded warning samples.
- Add focused ordinary and privileged ownership tests plus a Linux CI target.

## How to verify

- `make test:unit:guest`
- `make test:guest-perms`
- `make fmt:check:rust`
- `make clippy`
- `git diff --check`

## Risks / rollout

- Ownership repair now depends on Linux filesystem syscalls instead of GNU `chown`; privileged Linux tests cover recursive repair, read-only mounts, cross-mount traversal, and bind-mount cycles.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved guest ownership repair across files, directories, symlinks, and special files.
  * Ownership repair now continues safely when individual entries cannot be accessed.
  * Added protection against directory cycles and unsafe symlink traversal.
  * Added detailed repair results and warning reporting.

* **Tests**
  * Expanded coverage for permissions, ownership, filesystem errors, links, mounts, and edge cases.
  * Added automated privileged ownership tests on Linux.

### CODE REVIEW

**Review by DorianZheng — APPROVED** [review]

File: src/guest/src/storage/perms.rs [hunk 1]
```diff
@@ -97,3 +105,362 @@ impl OwnershipFixer {
...
+
+        if stat.st_mode & libc::S_IFMT != libc::S_IFDIR {
+            self.chown_operand(path);
+            return self.report;
+        }
+
+        // now it is directory, we need to traverse it
+        let flags = OFlag::O_RDONLY | OFlag::O_DIRECTORY | OFlag::O_NOFOLLOW | OFlag::O_CLOEXEC;
+        let root = match Dir::open(path, flags, Mode::empty()) {
+            Ok(root) => root,
```
[the remark this answers is not in the document]
Comment (ltstriker) [reply src/guest/src/storage/perms.rs:153]: this just follow the original logic. We just do the chown replace, do not change the main logic.

File: src/guest/src/storage/perms.rs [hunk 2]
```diff
@@ -97,3 +105,362 @@ impl OwnershipFixer {
...
+            None,
+            path,
+            Some(self.uid),
+            Some(self.gid),
+            AtFlags::AT_SYMLINK_NOFOLLOW,
+        ) {
+            Ok(()) => self.report.changed += 1,
+            Err(error) => self.report.record_failure("chown", path, error),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ltstriker) [reply src/guest/src/storage/perms.rs:331]: this just follow the original logic. We just do the chown replace, do not change the main logic.

File: src/guest/src/storage/perms/tests.rs [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+    let child = target.join("child");
+    fs::create_dir_all(&child).expect("create symlink target child");
+    let link = temp.path().join("link");
+    symlink(&target, &link).expect("create root symlink");
+
+    OwnershipFixer::fix_if_needed(&link)
+        .expect("legacy sampling follows a matching symlink operand");
+    OwnershipFixer::fix_if_needed(&link.join("child"))
+        .expect("legacy sampling follows a matching intermediate symlink");
+}
```
[the remark this answers is not in the document]
Comment (ltstriker) [reply src/guest/src/storage/perms/tests.rs:305]: this just follow the original logic. We just do the chown replace, do not change the main logic.

see:
> -P’
Do not traverse any symbolic links. This is the default if none of -H, -L, or -P is specified. See [Traversing symlinks](https://www.gnu.org/software/coreutils/manual/html_node/Traversing-symlinks.html).

File: src/guest/src/storage/perms/tests.rs [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+    );
+}
+
+#[test]
+fn handles_hardlinks_fifo_and_unix_socket() {
+    let target_dir = Path::new(env!("CARGO_MANIFEST_DIR")).join("../../target");
+    let temp = tempfile::Builder::new()
+        .prefix("boxlite-perms-")
+        .tempdir_in(target_dir)
+        .expect("create temporary directory");
```
[the remark this answers is not in the document]
Comment (ltstriker) [reply src/guest/src/storage/perms/tests.rs:230]: Good catch. The current CI uses Cargo’s default target directory and passes, but a clean checkout using a custom CARGO_TARGET_DIR can fail because tempdir_in does not create the hard-coded ../../target parent. This is test-only and does not affect production behavior, but using the configured target directory would make the test portable.
